### PR TITLE
IMPRO-1490: Wind vector to UV

### PR DESCRIPTION
**Pull request by neilCrosswaite**
# Added a cli to convert speed and direction to u and v.
As multiple CLI's are now using either `speed` and `direction` cubes or `u` and `v`, rather than error handling for all of them, this CLI converts from `speed` and `direction` to `u` and `v` so then the other CLIs can do the task they were designed for.

### Testing

Help and null are no longers being tested as they won't be once we move over to pytests.

It is not possible to test for the `TypeError` as part of the bats tests. This is because clize will warn you already as it is a required argument. The `raise TypeError` is still there for when it is being accessed via python. 

Testing:
 - [x] Ran tests and they passed OK
 - [x] Added new tests for the new feature(s)

### CODE REVIEW

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- lib/improver/cli/wind_vector_to_uv.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/improver/improver/pullRequest?prid=4569506&bid=15360774)

File: lib/improver/cli/__init__.py [hunk 1]
```diff
@@ -144,6 +145,27 @@ def inputcube(to_convert):
...
+        Loaded cube or passed object.
+
+    """
+    from improver.utilities.load import load_cube
+
+    speed_cube = maybe_coerce_with(load_cube, to_convert,
+                                   constraints="wind_speed")
+    direction_cube = maybe_coerce_with(load_cube, to_convert,
+                                       constraints="wind_from_direction")
+    return CubeList([speed_cube, direction_cube])
```
Comment (arh89) [inline lib/improver/cli/__init__.py]: I don't think we want to have something wind specific leaking into this file - we should try and keep it generic.

Maybe we could have something like a `create_constrainedinputcube_converter`, which is a higher order function which returns converter which applies a given constraint? (cc @TomekTrzeciak, who might hate this idea)

Something like...
```
def create_constrainedinputcube_converter(constraints):
    def constrainedinputcube_converter(to_convert):
        return maybe_coerce_with(load_cube, to_convert,
                                                   constraints=constraints)
    return constrainedinputcube_converter
```

Then in `wind_vector_to_uv.py`, you could have something like
```
inputspeedcube = cli.create_constrainedinputcube_converter('wind_speed')
inputdircube = cli.create_constrainedinputcube_converter('wind_from_direction')
def process(wind_speed: inputspeedcube,
            wind_direction: inputdircube):
```

File: lib/improver/cli/wind_vector_to_uv.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+"""Script to convert wind vector cubeList to a cubeList of U and V"""
+from iris import Constraint
+from iris.cube import CubeList
+
+from improver import cli
+from improver.wind_calculations.wind_components import ResolveWindComponents
```
Comment (arh89) [inline lib/improver/cli/wind_vector_to_uv.py]: In order to speed up the CLIs, we now put the `import`s (with the exception of `from improver import cli` into the `process` function.

File: lib/improver/cli/wind_vector_to_uv.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+"""Script to convert wind vector cubeList to a cubeList of U and V"""
```
Comment (arh89) [inline lib/improver/cli/wind_vector_to_uv.py]: Maybe something like
```suggestion
"""Convert wind speed and direction into individual velocity components."""
```

File: lib/improver/cli/wind_vector_to_uv.py (hunk 2 above)
Comment (neilCrosswaite) [reply lib/improver/cli/wind_vector_to_uv.py]: Fixed

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- lib/improver/cli/wind_vector_to_uv.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/improver/improver/pullRequest?prid=4569506&bid=15360774)

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- lib/improver/cli/wind_vector_to_uv.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/improver/improver/pullRequest?prid=4569506&bid=15360774)

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- lib/improver/cli/wind_vector_to_uv.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/improver/improver/pullRequest?prid=4569506&bid=15360774)

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- lib/improver/cli/wind_vector_to_uv.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/improver/improver/pullRequest?prid=4569506&bid=15360774)

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- lib/improver/cli/wind_vector_to_uv.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/improver/improver/pullRequest?prid=4569506&bid=15360774)

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- lib/improver/cli/wind_vector_to_uv.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/improver/improver/pullRequest?prid=4569506&bid=15360774)

File: lib/improver/cli/__init__.py [hunk 4]
```diff
@@ -220,6 +220,36 @@ def with_intermediate_output(wrapped, *args, intermediate_output=None,
     return result
 
 
+def create_constrained_inputcube_converter(*constraints):
+    """Makes a list of functions with different constraints.
+
+    Args:
+        *constraints (string):
```
Comment (neilCrosswaite) [inline lib/improver/cli/__init__.py]:
```suggestion
        *constraints (str):
```

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- lib/improver/cli/wind_vector_to_uv.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/improver/improver/pullRequest?prid=4569506&bid=15360774)

File: lib/improver/cli/resolve_wind_components.py [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+        wind_from_direction (iris.cube.Cube):
+            A cube of wind from direction.
+
+    Returns:
+        iris.cube.Cubelist:
+            A cubelist of the speed and direction as U and V cubes.
+    """
+    from iris.cube import CubeList
+    from improver.wind_calculations.wind_components import (
+        ResolveWindComponents)
```
Comment (arh89) [inline lib/improver/cli/resolve_wind_components.py]:
```suggestion
        ResolveWindComponents)

```

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- lib/improver/cli/resolve_wind_components.py  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/improver/improver/pullRequest?prid=4569506&bid=15360774)

File: lib/improver/cli/resolve_wind_components.py [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+    if wind_speed is None or wind_from_direction is None:
```
Comment (arh89) [inline lib/improver/cli/resolve_wind_components.py]:
```suggestion
if not (wind_speed and wind_from_direction):
```
? (but I don't really mind)

File: lib/improver/cli/resolve_wind_components.py [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+# POSSIBILITY OF SUCH DAMAGE.
+"""Convert wind speed and direction into individual velocity components."""
+
+from improver import cli
+
+
+@cli.clizefy
+@cli.with_output
+def process(wind_speed: cli.inputcube, wind_from_direction: cli.inputcube):
+    """Converts speed and direction individual velocity components.
```
Comment (gavinevans) [inline lib/improver/cli/resolve_wind_components.py]: Typo: `Converts speed and direction individual velocity components` => `Converts speed and direction into individual velocity components`

Comment (arh89) [inline lib/improver/cli/resolve_wind_components.py]:
```suggestion
    """Converts speed and direction into individual velocity components.
```
(for ease of commit) :)

Comment (neilCrosswaite) [reply lib/improver/cli/resolve_wind_components.py]: Fixed.

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- lib/improver/cli/resolve_wind_components.py  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/metoppv_tech/improver/pullRequest?prid=4569506&bid=15360774)

**Pull request merged**